backend/api/src/realms.rs

```diff
@@ -25,73 +38,126 @@ impl Realm {
+                    child_keys: vec![],
+                }
+            })
             .map_ok(|realm| (realm.key, realm))
             .try_collect::<HashMap<_, _>>().await?;
 
         // With this, and the `parent` member of the `Realm`,
         // we already have quick access to the data of a realm's parent.
-        // To also get to the children quickly we maintain another map.
-        let mut children = <HashMap<_, Vec<_>>>::new();
+        // To also get to the children quickly we maintain a corresponding list
+        // for each realm
+        let keys = realms.values()
+            .filter_map(
+                |realm| realm.parent_key.map(
+                    |parent_key| (realm.key, parent_key)
+                )
+            )
+            .collect::<Vec<_>>();
+        for (key, parent_key) in keys {
+            let parent = realms.get_mut(&parent_key).context("invalid parent")?;
+            parent.child_keys.push(key);
+        }
+
+        // After this point, we should know the tree structure to be valid.
+        // That is, we can now safely panic if we can't find things in our maps/lists;
+        // that's totally a bug in this code, then, not an inconsistency in the db.
+
+        // We also need a map from the full path to the proper realm,
+        // and conversely, we want to cache the full path inside the realm.
+        let from_path = index_by_path(&mut realms);
+
+        fn index_by_path(
+            realms: &mut HashMap<u64, Realm>,
+        ) -> HashMap<String, u64> {
 
-        for Realm { key, parent_key, .. } in realms.values() {
-            if key != parent_key {
-                children.entry(*parent_key).or_default().push(*key);
+            let mut index = HashMap::new();
+
+            fill_index_recursively(0, "", realms, &mut index);
+
+            fn fill_index_recursively(
+                parent: u64,
+                full_path: &str,
+                realms: &HashMap<u64, Realm>,
+                index: &mut HashMap<String, u64>,
+            ) {
+                index.insert(full_path.to_owned(), parent);
+
+                let parent = &realms[&parent];
+                for child_key in &parent.child_keys {
+                    let child = &realms[&child_key];
+                    let path = [full_path, &child.path_segment].join("/");
+
+                    fill_index_recursively(*child_key, &path, realms, index);
+                }
             }
+
+            index
         }
 
-        Ok(Tree { realms, children })
+        Ok(Tree { realms, from_path })
     }
 
     pub(crate) fn get_node(&self, id: &Id) -> Option<&Realm> {
         self.realms.get(&id.key_for(Id::REALM_KIND)?)
     }
 
     pub(crate) fn root(&self) -> &Realm {
-        self.realms.get(&0).expect("bug: no root realm")
+        &self.realms[&0]
+    }
+
+    pub(crate) fn from_path(&self, path: &str) -> Option<&Realm> {
+        self.realms.get(self.from_path.get(path)?)
```

Comment:
```suggestion
        self.from_path.get(path).map(|key| &self.realms[key])
```

I would say. The outer index operation should never fail.

Reply:
True.